### PR TITLE
chore: omega consume Expr metadata when required

### LINKED ISSUE #509: `omega` has issues with `mdata` handling
This MWE is [from Damiano Testa](https://leanprover.zulipchat.com/#narrow/stream/239415-metaprogramming-.2F-tactics/topic/A.20simple.20result.20that.20.60omega.60.20fails.20to.20prove/near/410999690), and I've hit it in the wild in [other cases](https://leanprover.zulipchat.com/#narrow/stream/239415-metaprogramming-.2F-tactics/topic/A.20simple.20result.20that.20.60omega.60.20fails.20to.20prove/near/411264731):

```
import Std.Tactic.Omega

example : 0 = 0 := by
  have : 0 = 0 := by omega -- works
  omega -- fails
```

He suggests it's `mdata`-related, and indeed a [tactic he wrote that clears `mdata`](https://leanprover.zulipchat.com/#narrow/stream/239415-metaprogramming-.2F-tactics/topic/A.20simple.20result.20that.20.60omega.60.20fails.20to.20prove/near/411009284) fixes the problem.

### PR DESCRIPTION
**Pull request by semorrison**
Closes #509.

### CODE REVIEW

**Comment by digama0** [comment]
Usually `whnfR` does what you want here...

**Comment by semorrison** [comment]
For now at least this function isn't in `MetaM`. (Although after some of the other concurrent PRs it will be.)

**Comment by digama0** [comment]
I see that. Do you want me to merge this as is and repeat my suggestion on the other PRs, or change it to MetaM here and deal with the merge conflict?

**Comment by semorrison** [comment]
I'll change it to `MetaM` here, and deal with the merge conflicts in whatever order they merge!

File: Std/Tactic/Omega/Frontend.lean [hunk 1]
```diff
@@ -246,29 +246,39 @@ def addIntInequality (p : MetaProblem) (h y : Expr) : OmegaM MetaProblem := do
...
+  | (``LE.le, #[.const ``Nat [], _, x, y]) =>
+    return some (mkApp3 (.const ``Nat.lt_of_not_le []) x y h)
+  | (``GT.gt, #[.const ``Int [], _, x, y]) =>
+    return some (mkApp3 (.const ``Int.le_of_not_lt []) y x h)
+  | (``GE.ge, #[.const ``Int [], _, x, y]) =>
+    return some (mkApp3 (.const ``Int.lt_of_not_le []) y x h)
+  | (``GT.gt, #[.const ``Nat [], _, x, y]) =>
+    return some (mkApp3 (.const ``Nat.le_of_not_lt []) y x h)
+  | (``GE.ge, #[.const ``Nat [], _, x, y]) =>
+    return some (mkApp3 (.const ``Nat.lt_of_not_le []) y x h)
```
Comment (digama0) [inline Std/Tactic/Omega/Frontend.lean:266]:
```suggestion
```
with `whnfR` these have probably become unreachable

**Comment by joehendrix** [comment]
Maybe this is largely orthogonal, but came up with the call to `whnfR`.  Should `omega` generally see through reducible definitions?  e.g.,

```lean
@[reducible] def z : Nat := 0

example : z = 0 := by
  omega
```

**Comment by mattrobball** [comment]
This came up in leanprover-community/mathlib#9594 where I tried to swap out `linarith` for `omega` but had to fall back to chaining them.

**Comment by semorrison** [comment]
> Maybe this is largely orthogonal, but came up with the call to `whnfR`. Should `omega` generally see through reducible definitions? 

This does seem desirable, but I'd prefer to do it in a separate PR so I can look at the performance first.

**Comment by semorrison** [comment]
I've just merged this into #518, so will close this.

**Pull request closed without merging**